datanode/src/main/java/com/odysseusinc/arachne/datanode/datasource/CheckDataSourceService.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2018, 2023 Odysseus Data Services, Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.odysseusinc.arachne.datanode.datasource;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.odysseusinc.arachne.commons.api.v1.dto.CommonModelType;
+import com.odysseusinc.arachne.datanode.exception.ValidationException;
+import com.odysseusinc.arachne.datanode.model.datasource.DataSource;
+import com.odysseusinc.arachne.datanode.service.ExecutionEngineIntegrationService;
+import com.odysseusinc.arachne.datanode.util.SqlUtils;
+import com.odysseusinc.arachne.execution_engine_common.api.v1.dto.AnalysisRequestDTO;
+import com.odysseusinc.arachne.execution_engine_common.api.v1.dto.AnalysisRequestStatusDTO;
+import com.odysseusinc.arachne.execution_engine_common.api.v1.dto.AnalysisRequestTypeDTO;
+import com.odysseusinc.arachne.execution_engine_common.api.v1.dto.AnalysisResultDTO;
+import com.odysseusinc.arachne.execution_engine_common.api.v1.dto.DataSourceUnsecuredDTO;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.ArrayUtils;
+import org.ohdsi.sql.SqlTranslate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import static com.odysseusinc.arachne.datanode.datasource.CheckDataSourceCallbackController.CHECK_RESULT;
+import static com.odysseusinc.arachne.datanode.datasource.CheckDataSourceCallbackController.CHECK_UPDATE;
+import static com.odysseusinc.arachne.execution_engine_common.api.v1.dto.AnalysisResultStatusDTO.EXECUTED;
+
+@Slf4j
+@Service
+@Transactional
+public class CheckDataSourceService {
+    private static final String CDM_VERSION_FILENAME = "cdm_version.txt";
+
+    private final Cache<Long, CompletableFuture<CheckResult>> results = CacheBuilder.newBuilder().expireAfterAccess(60, TimeUnit.MINUTES).build();
```

Comment:
How often do we check the database? Is this a hard bit? If so, does that mean we only check the availability of the data source once since we access the cache more frequently than once per hour?

Reply:
We check every minute. The hour here is mostly to have comfort when debugging.